### ISSUE TITLE
Date Shows 

### ISSUE THREAD

**Issue by nuvious** [jupyterlab/jupyterlab#16254]
## Description

Last Modified date shows "next year" for new notebooks and existing notebooks soon after being saved, especially while
under a heavy processing load.

## Reproduce

1. Create a new notebook:

![image](https://github.com/jupyterlab/jupyterlab/assets/5287736/d8ef7f56-c381-4ef2-8373-da22c47d3048)

2. Click on File > Save Notebook/Press CTRL+S/Click Save Button and save the notebook

![image](https://github.com/jupyterlab/jupyterlab/assets/5287736/568ea3c7-ba1d-414d-bb8e-14c5ea9a0fd0)

3. Add some code that will take a long time to process, run it, and then click save:

![before](https://github.com/jupyterlab/jupyterlab/assets/5287736/8db38ddf-6134-4223-92df-2c77ebf2dcba)

![after clicking save](https://github.com/jupyterlab/jupyterlab/assets/5287736/7a8b2c81-8188-470f-a226-69b02622d128)

Diagnosed and narrowed the issue down to `Time.formatHuman` function in
`jupyter/jupyterlab/packages/coreutils/src/time.ts`. If you set a breakpoint after `delta` is calculated it is sometimes
a positive value. My guess is this is due to some asynchronous execution resulting in the save time appearing before the
Date.now() time.

![image](https://github.com/jupyterlab/jupyterlab/assets/5287736/5b0013be-6cf4-4732-a34a-cd8f33b4135d)

With `delta` (`s`) being a positive value, the result of `Math.ceil(delta / unit.milliseconds)` for any of the units
will be 1. The first unit in the list is `{ name: 'years', milliseconds: 365 * 24 * 60 * 60 * 1000 }` and thus with
amount being non-zero `return formatter.format(amount, unit.name);` on line 57 has the values 1, "year" and is formatted
to "next year".

A simple check to see if s is positive and return `formatter.format(0, 'seconds');` if it is will resolve the issue and
a PR is to follow submitted by me; just wanted to get the issue documented first.

## Expected behavior

Expect it to show "0 seconds" or "now" or whatever makes sense based on notebook language settings.

## Context

- Operating System and version: Debian GNU/Linux 12 (bookworm)
- Browser and version:
  - Firefox 125.0.2 (64-bit)
  - Edge 124.0.2478.51 (Official build) (64-bit)
- JupyterLab version: 4.1.5

<details><summary>Troubleshoot Output</summary>
<pre>
$PATH:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin
        /home/nuvious/.vscode-server/cli/servers/Stable-e170252f762678dec6ca2cc69aba1570769a5d39/server/bin/remote-cli
        /usr/local/bin
        /usr/bin
        /bin
        /usr/local/games
        /usr/games

sys.path:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin
        /usr/lib/python311.zip
        /usr/lib/python3.11
        /usr/lib/python3.11/lib-dynload
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/lib/python3.11/site-packages

sys.executable:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin/python3

sys.version:
        3.11.2 (main, Mar 13 2023, 12:18:29) [GCC 12.2.0]

platform.platform():
        Linux-6.1.0-18-amd64-x86_64-with-glibc2.36

which -a jupyter:
        /home/nuvious/EN.695.737.81.SP24-ArtificialIntelligenceForAssuredAutonomy/.venv/bin/jupyter

pip list:
        Package                      Version
        ---------------------------- --------------
        absl-py                      2.1.0
        amalgam-lang                 10.0.0
        anyio                        4.3.0
        argon2-cffi                  23.1.0
        argon2-cffi-bindings         21.2.0
        arrow                        1.3.0
        asttokens                    2.4.1
        astunparse                   1.6.3
        async-lru                    2.0.4
        attrs                        23.2.0
        Babel                        2.14.0
        bayesian-optimization        1.4.3
        beautifulsoup4               4.12.3
        bleach                       6.1.0
        certifi                      2024.2.2
        cffi                         1.16.0
        charset-normalizer           3.3.2
        cloudpickle                  3.0.0
        colorama                     0.4.6
        comm                         0.2.2
        contourpy                    1.2.0
        cryptography                 42.0.5
        cycler                       0.12.1
        debugpy                      1.8.1
        decorator                    5.1.1
        defusedxml                   0.7.1
        deprecation                  2.1.0
        executing                    2.0.1
        Faker                        24.3.0
        fastjsonschema               2.19.1
        flatbuffers                  24.3.7
        fonttools                    4.50.0
        fqdn                         1.5.1
        gast                         0.5.4
        google-pasta                 0.2.0
        grpcio                       1.62.1
        h11                          0.14.0
        h5py                         3.10.0
        howso-engine                 19.0.0
        howso-openapi-client         32.0.0
        httpcore                     1.0.4
        httpx                        0.27.0
        humanize                     4.9.0
        idna                         3.6
        imageio                      2.34.1
        ipykernel                    6.29.3
        ipython                      8.22.2
        ipywidgets                   8.1.2
        isoduration                  20.11.0
        jedi                         0.19.1
        Jinja2                       3.1.3
        joblib                       1.3.2
        json5                        0.9.24
        jsonpointer                  2.4
        jsonschema                   4.21.1
        jsonschema-specifications    2023.12.1
        jupyter                      1.0.0
        jupyter_client               8.6.1
        jupyter-console              6.6.3
        jupyter_core                 5.7.2
        jupyter-events               0.10.0
        jupyter-lsp                  2.2.4
        jupyter_server               2.13.0
        jupyter_server_terminals     0.5.3
        jupyterlab                   4.1.5
        jupyterlab_pygments          0.3.0
        jupyterlab_server            2.25.4
        jupyterlab_widgets           3.0.10
        keras                        3.1.1
        kiwisolver                   1.4.5
        lazy_loader                  0.4
        libclang                     18.1.1
        lime                         0.2.0.1
        llvmlite                     0.42.0
        Markdown                     3.6
        markdown-it-py               3.0.0
        MarkupSafe                   2.1.5
        matplotlib                   3.8.3
        matplotlib-inline            0.1.6
        mdurl                        0.1.2
        mistune                      3.0.2
        ml-dtypes                    0.3.2
        mmh3                         4.1.0
        namex                        0.0.7
        nbclient                     0.10.0
        nbconvert                    7.16.3
        nbformat                     5.10.3
        nest-asyncio                 1.6.0
        networkx                     3.3
        notebook                     7.1.2
        notebook_shim                0.2.4
        numba                        0.59.1
        numpy                        1.26.4
        opt-einsum                   3.3.0
        optree                       0.10.0
        overrides                    7.7.0
        packaging                    24.0
        pandas                       2.2.1
        pandocfilters                1.5.1
        parso                        0.8.3
        pexpect                      4.9.0
        pillow                       10.2.0
        pip                          23.0.1
        platformdirs                 4.2.0
        prometheus_client            0.20.0
        prompt-toolkit               3.0.43
        protobuf                     4.25.3
        psutil                       5.9.8
        ptyprocess                   0.7.0
        pure-eval                    0.2.2
        pycparser                    2.21
        pygad                        3.3.1
        Pygments                     2.17.2
        PyJWT                        2.8.0
        pyparsing                    3.1.2
        python-dateutil              2.9.0.post0
        python-json-logger           2.0.7
        pytz                         2024.1
        PyYAML                       6.0.1
        pyzmq                        25.1.2
        qtconsole                    5.5.1
        QtPy                         2.4.1
        referencing                  0.34.0
        requests                     2.31.0
        rfc3339-validator            0.1.4
        rfc3986-validator            0.1.1
        rich                         13.7.1
        rpds-py                      0.18.0
        scikit-image                 0.23.2
        scikit-learn                 1.4.1.post1
        scipy                        1.12.0
        semantic-version             2.10.0
        Send2Trash                   1.8.2
        setuptools                   66.1.1
        shap                         0.45.0
        six                          1.16.0
        slicer                       0.0.7
        sniffio                      1.3.1
        soupsieve                    2.5
        stack-data                   0.6.3
        tensorboard                  2.16.2
        tensorboard-data-server      0.7.2
        tensorflow                   2.16.1
        tensorflow-io-gcs-filesystem 0.36.0
        termcolor                    2.4.0
        terminado                    0.18.1
        threadpoolctl                3.4.0
        tifffile                     2024.4.24
        tinycss2                     1.2.1
        tornado                      6.4
        tqdm                         4.66.2
        traitlets                    5.14.2
        types-python-dateutil        2.9.0.20240316
        typing_extensions            4.10.0
        tzdata                       2024.1
        uri-template                 1.3.0
        urllib3                      2.2.1
        wcwidth                      0.2.13
        webcolors                    1.13
        webencodings                 0.5.1
        websocket-client             1.7.0
        Werkzeug                     3.0.1
        wheel                        0.43.0
        widgetsnbextension           4.0.10
        wrapt                        1.16.0
</pre>
</details>

<details><summary>Command Line Output</summary>
<pre>
N/A - Does not throw an exception and source of issue was isolated with javascript/typescript debugging and static
analysis.
</pre>
</details>

<details><summary>Browser Output</summary>

<pre>
See screenshot for specific issue found in TypeScript
</pre>
</details>

**Comment by paloha** [comment]
I have a similar/same problem. I am running Jupyter Lab on a server. I connect to it from my local machine via an ssh tunnel. I checked by `watch -n 1 date` that the time appears to be the same for my machine and the server. To reproduce, I just need to create a a new notebook or save it. It always shows next year. When I hover over the "next year" string, it says it was created approx. 1 minute in the future w.r.t. the server clock.

**Comment by nuvious** [comment]
@paloha, I opened a PR on this a while back and decided to dust it off. Got it working again in the devcontainer but there are two tests failing right now; one of which seems to just be a screenshot test. Hope to work with the reviewers to get the PR polished to their liking.

**Comment by jon-abbott** [comment]
I'm also experiencing this issue on JupyterLab 4.4.1.

**Comment by janxkoci** [comment]
I see this in jupyterlab 4.4.5 (I should update, I know, I'm busy) with a file that comes from a Debian server that doesn't sync time with the internet (apparently that's not done [by default](https://wiki.debian.org/NTP)) so the server's time is actually some 30 minutes in the future. The timestamp for that file in the jupyterlab file viewer shows as "next yr.":

<img width="560" height="253" alt="Image" src="https://github.com/user-attachments/assets/7df067c5-7ae6-48d6-8ed0-2b70483a9c61" />

**Comment by krassowski** [comment]
On some level I would think making the servers use the correct time is the solution (and a responsibility of server maintainers, not Jupyter), but since this appears to bother so many people, do we need to expose the server time in the `ContentManager` response and use that as a reference?

**Comment by janxkoci** [comment]
I agree about the server settings etc, my post was more of a hint at situations when this issue may appear. I suspect something like "current minute" may appear as "next yr." too. (Not to mention "next year" is definitely not coming anywhere near that soon.)

PS: about this:

> this appears to bother so many people

It doesn't really bother me, I've found it rather amusing. I just thought this was some js framework to blame 🤷 But I think I care more about fixing our server than this issue here, but I cannot speak for others..

**Comment by krassowski** [comment]
Back-linking discussion on Zulip re Binder time being out of sync: [#jupyterhub > What needs to be done to make Binder clock run correctly](https://jupyter.zulipchat.com/#narrow/channel/469744-jupyterhub/topic/What.20needs.20to.20be.20done.20to.20make.20Binder.20clock.20run.20correctly/with/546541668)

**Comment by xhejtman** [comment]
Most probably it is just race condition in code that can rarely happen, if the jupyter code is like:

```
now = time();
<----- now the notebook is saved
stat(notebookfilename, &sb);

last_modified = now - sb.st_mtim.tv_sec;
```

it can be negative when the notebook is saved in the exactly right time. If should be fixed so that `if abs(now - sb.st_mtim.tv_sec) < 2`, set last modified to now.
